containers-workarounds.md

```diff
@@ -11,13 +11,98 @@ We've documented ways to leverage these products below:
+wget https://golang.org/dl/go1.16.5.linux-arm64.tar.gz &&
+sudo tar -C /usr/local/ -xzf go1.16.5.linux-arm64.tar.gz &&
+sudo ln -s /usr/local/go/bin/go /usr/local/bin/go
+```
 
+2. Logout and log back in after the previous step. Optionally reboot to apply any kernel updates that were downloaded.
 
+3. Clone the repositories.
+```
+cd ~
+mkdir istio && cd istio
+git clone https://github.com/istio/istio.git
+git clone https://github.com/istio/proxy.git
+git clone https://github.com/istio/tools.git
+```
 
+3.1. Optionally checkout the exact versions tested for this guide. If you run into problems, try this first. (Note: these commands will reset the master branch, so do not execute them if you have your own unpublished work on the master branch.)
+```
+cd istio && git checkout -B master 49dfd655557fa5e1f678d6c18761603ceddcb1f8 && cd ..
+cd proxy && git checkout -B master b3ba3a91e80939fbbb65ec998d812b422b3afefb && cd ..
+cd tools && git checkout -B master 236cb4951368852f6fdc568ce7ffdc3291d6ac4d && cd ..
+```
 
+4. Build the "build" containers. This step took about 15 minutes to run on a `m6g.4xlarge`. It does not use much parallelism so a larger instance is not likely to improve the speed.
+```
+cd ~/istio/tools/docker/build-tools
+DRY_RUN=1 time ./build-and-push.sh
+```
 
+5. Build the proxy (envoy) binary. This step takes about 20 minutes on a `m6g.16xlarge` and makes full use of all cores. A smaller instance will run for longer.
+```
+docker run --name istio-proxy-build -it -v $(realpath ~/istio/proxy/):/work -w /work -u $(id -u) gcr.io/istio-testing/build-tools-proxy:master-latest make build
+```
+
+6. Build the Go binaries. This step runs in just a minute or two.
+```
+cd ~/istio/istio
+TARGET_ARCH=arm64 IMAGE_VERSION=master-latest make build
+```
+
+7. Build the final containers. For this step we set several environment variables, so start by entering a new shell to contain the new environment.
+```
+bash
+export IMAGE_TAG=master-latest
+export IMAGE_VERSION=$IMAGE_TAG
+export DOCKER_ARCHITECTURES="linux/arm64"
+export TARGETARCH="arm64"
+export HUBS="gcr.io/istio-release"
+export TAG=$IMAGE_TAG
+export BASE_VERSION=$IMAGE_TAG
+cd ~/istio/proxy
+export PROXY_REPO_SHA=$(git rev-parse HEAD)
+cd ..
+```
+
+7.1. Copy the `envoy` binary into place.
+```
+docker cp istio-proxy-build:/work/bazel-bin/src/envoy/envoy ~/istio/istio/out/linux_arm64/release/envoy-${PROXY_REPO_SHA}
+cp ~/istio/istio/out/linux_arm64/release/envoy-${PROXY_REPO_SHA} ~/istio/istio/out/linux_arm64/release/envoy
+```
+
+7.2. Build the base containers.
+```
+cd ~/istio/istio
+for i in 'dockerx.base' 'dockerx.distroless' 'dockerx.app_sidecar_base_ubuntu_xenial' 'dockerx.app_sidecar_base_ubuntu_bionic' 'dockerx.app_sidecar_base_ubuntu_focal' 'dockerx.app_sidecar_base_debian_9' 'dockerx.app_sidecar_base_debian_10' 'dockerx.app_sidecar_base_centos_8' 'dockerx.app_sidecar_base_centos_7'; do HUBS="gcr.io/istio-release" make $i; done
```

Comment:
you're only building the arm64 containers here? why is buildx needed?

Reply:
There are many parallel build steps that buildx does execute on the next step. It's really not supposed to be necessary to run all these individually because the project uses a "bakefile" to build all of them at once, but for some reason it can't get the dependency ordering right and it fails. Also the dockerfiles use `$TARGETARCH` which is automatically set by buildx. Istio can build without buildx, but I didn't get it to work correctly.